alphadia/workflow/optimizer.py

```diff
@@ -0,0 +1,147 @@
+# native imports
+from functools import wraps
+
+import numpy as np
+
+# alpha family imports
+# third party imports
+import pandas as pd
+
+# alphadia imports
+from alphadia.workflow import reporting
+
+
+def check_if_optimum_found(method):
+    """If an optimum has been found, the optimizer should cease to execute its methods to check, initiate and update.
+    This function can be used as a decorator and placed before the methods in an optimizer to prevent execution of the methods if an optimum has been found.
+    If the optimum has been found, the decorator will print the end of optimization message and, for consistency with stop_or_continue methods, return True.
+    """
+
+    @wraps(method)
+    def wrapper(self, *args, **kwargs):
+        if self.optimal_tolerance is not None:
+            self.end_of_optimization_message()
+            return True
```

Comment:
I feel this is a bit of a dangerous case of how to use decorators because you don't know if the method you are wrapping is actually returning a boolean or something else ..
 .. still, I second of course your striving for less code duplication :-)

Think about this alternative:

add this stopping logic into `BaseOptimizer`:
```python
def update(self):
        if self.optimal_tolerance is not None:
            self.end_of_optimization_message()
            return
         self._update()

def _update(self):
        raise NotImplementedError(
            f"_update() not implemented for {self.__class__.__name__}"
        )

# subclasses implement `_update`
```
This way, you have the stopping logic centralized (although with a bit more code dup).

Reply:
after reading this comment https://github.com/MannLabs/alphadia/pull/283/files#r1691005405 :
 could also be something like
```python
def update(self) -> bool:
# return value indicates wheter to stop
         self._update()
         return check_stopping_conditions()

def _update(self):
...
```

Reply:
With respect to the decorator, what do you think of Georg's suggestion(https://github.com/MannLabs/alphadia/pull/283#discussion_r1690983718)? Albetít I think that if I combine the check_stopping_conditions and initiate methods as suggested elsewhere then the decorator is probably superfluous. 

I agree setting the return value is a bad idea.

Reply:
I think the decorator will cause more trouble than good in the long run ;-)